### ISSUE TITLE
Make sure the field for cell count in the project form is linked to the metadata schema

### ISSUE THREAD

**Issue by gabsie** [ebi-ait/dcp-ingest-central#425]
As a data wrangler I want to be able to enter the cell count for a project, so that we can prioritise these projects better, visualise that information in the DCP ingest backoffice and the HCA DCP browser.

Acceptance Criteria / Definition of Done

Given I am entering details about a project
When I see a field for cell count and enter the number
Then this information will be saved in ingest, as well as link to metadata schema and visualise eventually in the DCP browser.

 

- [ ] Linking to the metadata schema

**Comment by aaclan-ebi** [comment]
From what I understand, what we want here is that when a user updates a "cell count" field in the Ingest UI, this would appear as the `estimated_cell_count` property in the project metadata JSON that we are submitting to the HCA DCP browser.

The implementation would involve the following tasks:
 
1. Update existing projects to move the current `cellCount` field value to the `estimated_cell_count` property project metadata. This would require the existing projects to be migrated (update `describedBy` field) to use the latest project metadata schema version. These changes would need to be imported by the Terra Data Import team so that it can be propagated to DCP.

2.  Update the project catalogue to read from the `estimated_cell_count` field in the project metadata.

3. Deprecate the current `cellCount` field which is stored in Ingest but is not part of the project metadata schema. This requires task no. 1 & 2 to be completed first.

4. Update the Ingest UI so that the form will update the `estimated_cell_count` in the project metadata.

**Comment by amnonkhen** [comment]
related to ebi-ait/dcp-ingest-central#432 - this one is  about accession, but it explains the principal of migrating project data to a newer schema.

**Comment by aaclan-ebi** [comment]
Thanks @amnonkhen , i'll work on how to migrate the existing projects to use the latest schema version and update their cell count property.

**Comment by aaclan-ebi** [comment]
I worked on a script to migrate projects with cell count in Ingest - will have it ready for PR today.

The latest project schema version is **15.0.0** however, this is a breaking change as all projects with publications will need to be updated with boolean value for `official_hca_publication`.

The schema version **14.3.0** is the one that introduced the cell count field. It should be safe to migrate the fields to this version.

From running the script locally with a copy of prod DB, there are:

**411** projects to be updated with cell counts
**559** projects to update with newer schema version 14.3.0 in ingest
**189** projects which have submissions and some/all of them will have to be exported to propagate changes to DCP

I'm seeing a possible performance issue in exporting 189 submissions. Ingest Core API may not be able to handle the requests from the exporter. We should think of a way to optimise the exporter. Currently, the whole submission is being processed every time an update (to one or many entities ) is applied. For the migration of project cell counts, only the project metadata is going to be updated for 559 projects, but the exporter will try to crawl the assay subgraphs for ~189 projects.

I scheduled a meeting with @amnonkhen today to discuss how to approach this possible performance issue.

**Comment by aaclan-ebi** [comment]
Per discussion with @amnonkhen ,
1. We should do a run-through of the migration first in the dev/staging environment to see how the core will handle the exporting of many submission envelopes. 
2. It would be nice to track the last modified timestamp and user of the metadata update (check if this can be easily done)
3. Include the steps/tasks listed ebi-ait/dcp-ingest-central#432  in the migration needed for this ticket. Create separate tickets if necessary.
4. Confirm with wranglers if we only want to upgrade the existing project versions to 14.3.0 for the cell count. Do we want to collate all the migration changes and export everything in one go? How would we update the required fields?

**Comment by aaclan-ebi** [comment]
There are 9 projects which couldn't be updated due to the following error:

```
<ExceptionInfo><url>http://localhost:8080/projects/5cda9df1d96dad0008579758</url><exceptionMessage>Infinite recursion (StackOverflowError) (through reference chain: org.humancellatlas.ingest.project.Project$$EnhancerByCGLIB$$42b6ca24["supplementaryFiles"]->com.sun.proxy.$Proxy199[0]->org.humancellatlas.ingest.file.File["project"]->org.humancellatlas.ingest.project.Project$$EnhancerByCGLIB$$42b6ca24["supplementaryFiles"]->com.sun.proxy.$Proxy199[0]->org.humancellatlas.ingest.file.File["project"]->org.humancellatlas.ingest.project.Project$$EnhancerByCGLIB$$42b6ca24["supplementaryFiles"]-
...<truncated>
>com.sun.proxy.$Proxy199[0]->org.humancellatlas.ingest.file.File["project"]->org.humancellatlas.ingest.project.Project$$EnhancerByCGLIB$$42b6ca24["supplementaryFiles"]->com.sun.proxy.$Proxy199[0]->org.humancellatlas.ingest.file.File["submissionEnvelope"]->org.humancellatlas.ingest.submission.SubmissionEnvelope$$EnhancerByCGLIB$$34f75aee["target"]->org.humancellatlas.ingest.submission.SubmissionEnvelope["stagingDetails"]->org.humancellatlas.ingest.submission.StagingDetails["stagingAreaUuid"]->org.humancellatlas.ingest.core.Uuid["uuid"])</exceptionMessage></ExceptionInfo>
```

I need to investigate this.

**Comment by aaclan-ebi** [comment]
Update for today:
- PR created for update script last friday
- 9 projects can't be updated and I need to invest core error further, but will prioritise ingest-ui and project catalogue
- detailed list of tasks needed for this work
- Setting up a meeting today with @gabsie to discuss the migration and related changes to this ticket


Listing the questions/tasks from ebi-ait/dcp-ingest-central#432

> 1. Which projects need to be migrated? All or a subset?

All projects will be updated

> 2. how to validate modified projects?

If we update the project's content in Ingest (which contains the project metadata JSON) thru Ingest API, the metadata will be set back to Draft and should trigger the metadata validation.

> 3. Do we change projects' status?

I'm not sure what this means, but when metadata has been updated, the linked submission envelopes will be automatically set from Exported to Valid and it should be possible to submit them again for exporting to Terra.

> 4. what would the effects on exporting to Terra be?

Terra will just import the latest version of the project metadata file. 

> 5. migration script should include mandatory changes from interim schema changes (e.g. hca_publication in 15.0.0)

We could include this when we update the cell count. But we need input from wranglers on which projects should have hca_publication set to true. 

> project catalog updates. see ticket: XYZ 

ebi-ait/dcp-ingest-central#445

> 7. ingest-ui updates. see ticket: XYZ

i believe it's this ticket

**Comment by aaclan-ebi** [comment]
I'm moving all the migration work needed for cell count to ebi-ait/dcp-ingest-central#445 .

**Comment by aaclan-ebi** [comment]
I created PR for UI changes yesterday. I'll apply the review comments today.

**Comment by aaclan-ebi** [comment]
Merged PR, deployed to dev and staging:
![Screenshot 2021-09-15 at 15.12.43.png](https://images.zenhubusercontent.com/5b042c754b5806bc2bd426e2/d1951d36-059e-43bf-8294-1b70365627ed)

**Comment by aaclan-ebi** [comment]
Hi @Wkt8 , this should be ready for wrangler testing and can be tested alongside the project catalogue changes ebi-ait/dcp-ingest-central#446. 

For the test plan, I imagine test cases & expected behaviour will be:
1. **Create a project** - the project metadata created should have estimated_cell_count property with value the same as what is set in the UI form. The project catalogue should display the updated value.
2. **Update a project**
2.a **Update a project with a version before 14.3.0** (the version which introduced the estimated_cell_count) - it would not be possible to update the estimated_cell_count yet until we migrate the projects to use 14.3.0 or a later version. The form that will be rendered wouldn't contain any cell count field. This should wait for ticket ebi-ait/dcp-ingest-central#445
2.b **Update project with version after 14.3.0** - the project metadata updated should have estimated_cell_count property with value the same as what is set in the UI form. The project catalogue should display the updated value.

You can test in Ingest UI staging https://staging.contribute.data.humancellatlas.org/projects and the data there is being read by Project Catalogue dev instance https://wwwdev.ebi.ac.uk/humancellatlas/project-catalogue/ .

Thank you!

**Comment by aaclan-ebi** [comment]
screenshot:
![Screenshot 2021-09-14 at 09.00.14.png](https://images.zenhubusercontent.com/5b042c754b5806bc2bd426e2/1ac1773e-718b-40ed-9dc2-8c034a377d26)

**Comment by Wkt8** [comment]
Have tested the following test plans in staging: (skip to summary at the bottom)
1. Create a project - the project metadata created should have estimated_cell_count property with value the same as what is set in the UI form. The project catalogue should display the updated value.
	1. Testing creating a project via submission that does NOT have the estimated_cell_count property 
		1. Edit the estimated_cell_count property in the UI
		2. Download spreadsheet endpoint to check if it is in the updated spreadsheet (Works!)
		3. Test the project catalogue is updated (Works!)
		4. Edit the estimated_cell_count property by downloading and editing spreadsheet from spreadsheet endpoint
		5. Update using 'bulk updates'
		6. Check in the UI to see if it has updated (Works!)
		7. Test the project catalogue is updated (works!)
	2. Testing creating a project via submission that DOES have the estimated_cell_count_property
		1. Check in the UI (works)
		2. Test the project catalogue is updated (works)
	3. Testing creating a project via 'Register a Project' functionality with estimated_cell_count
		1. Check in the UI (works!) 
		2. Check spreadsheet endpoint (N/A as no submission) 
		3. Check catalogue (works)

2. Update a project
	1. Update a project with a version before 14.3.0 (the version which introduced the estimated_cell_count) - it would not be possible to update the estimated_cell_count yet until we migrate the projects to use 14.3.0 or a later version. The form that will be rendered wouldn't contain any cell count field. This should wait for ticket Migrate existing projects to use latest project metadata schema version 15.0.0 #445
		1. https://staging.contribute.data.humancellatlas.org/projects/detail?uuid=1c6a960d-52ac-44ea-b728-a59c7ab9dc8e 
		2. "describedBy" : "https://schema.humancellatlas.org/type/project/14.1.0/project",
		3. Attempted to update in the UI (no field for estimated_cell_count)
		4. Attempted to update by downloading the spreadsheet from the spreadsheet endpoint and adding the project.estimated_cell_count field with a value
			1. Results in adding the value to the project information UI tab 
			2. Results in adding the value to the project catalogue
			3. Results in expected error: should have required property 'official_hca_publication' at .publications[0]
	2. Update project with version after 14.3.0 - the project metadata updated should have estimated_cell_count property with value the same as what is set in the UI form. The project catalogue should display the updated value.
		1. Works


In Summary: The feature works as expected. The only issue in testing was the expected issue of attempting to update a project with a version before 14.3.0 using the 'bulk updates' feature. This resulted in the estimated_cell_count being added to the project (due to bulk updates currently automatically updating the schema versions which is being dealt with in a separate ticket) and the required official_hca_publication property.

**Comment by aaclan-ebi** [comment]
Thanks @Wkt8 for testing and noting the detailed steps! I will promote the UI and project catalogue changes to prod now!

**Comment by aaclan-ebi** [comment]
This is now in prod.

![Screenshot 2021-09-17 at 16 23 35](https://user-images.githubusercontent.com/31281398/133808243-c25236fe-e2ff-4443-b9af-7988fc02438a.png)
